include/real/real_math.hpp

```diff
@@ -0,0 +1,237 @@
+				
+				for(exact_number<T> i = (two * term_number) + one ; i <= two * (term_number + one); i += one){
+					factorial *= i;
+				}
+				cur_power *= square_x;
+				cur_term = cur_power;
+				cur_term.divide_vector(factorial, max_error_exponent, upper);
+				++ term_number_int;
+				term_number += one;
+				
+			}while(cur_term.abs() > max_error);
+			result = result.up_to(max_error_exponent, upper);
+			return result;
+		}
+
+		/**
+		 *  TANGENT FUNCTION USING TAYLOR EXPANSION
+		 * @brief: calculates tan(x) of a exact_number using taylor expansion
+		 * @param: x: the exact_number.
+		 * @param: max_error_exponent: Absolute Error in the result should be < 1*base^(-max_error_exponent)
+		 * @param:  upper: if true: error lies in [0, +epsilon]
+		 *                  else: error lies in [-epsilon, 0], here epsilon = 1*base^(-max_error_exponent)
+		 * @author: Vikram Singh Chundawat
+		 **/
+		template<typename T>
+		inline exact_number<T> tangent(exact_number<T> x, size_t max_error_exponent, bool upper){
+			exact_number<T> result = sine(x, max_error_exponent, upper);
+			exact_number<T> cos = cosine(x,  max_error_exponent, upper);
+			result.divide_vector(cos, max_error_exponent, upper);
+			result = result.up_to(max_error_exponent, upper);
+			return result; 
+		}
+
+		/**
+		 *  COTANGENT FUNCTION USING TAYLOR EXPANSION
+		 * @brief: calculates cot(x) of a exact_number using taylor expansion
+		 * @param: x: the exact_number.
+		 * @param: max_error_exponent: Absolute Error in the result should be < 1*base^(-max_error_exponent)
+		 * @param:  upper: if true: error lies in [0, +epsilon]
+		 *                  else: error lies in [-epsilon, 0], here epsilon = 1*base^(-max_error_exponent)
+		 * @author: Vikram Singh Chundawat
+		 **/
+		template<typename T>
+		inline exact_number<T> cotangent(exact_number<T> x, size_t max_error_exponent, bool upper){
+			exact_number<T> result = cosine(x, max_error_exponent, upper);
+			exact_number<T> sin = sine(x,  max_error_exponent, upper);
+			result.divide_vector(sin, max_error_exponent, upper);
+			result = result.up_to(max_error_exponent, upper);
+			return result; 
+		}
+
+		/**
+		 *  SECANT FUNCTION USING TAYLOR EXPANSION
+		 * @brief: calculates sec(x) of a exact_number using taylor expansion
+		 * @param: x: the exact_number.
+		 * @param: max_error_exponent: Absolute Error in the result should be < 1*base^(-max_error_exponent)
+		 * @param:  upper: if true: error lies in [0, +epsilon]
+		 *                  else: error lies in [-epsilon, 0], here epsilon = 1*base^(-max_error_exponent)
+		 * @author: Vikram Singh Chundawat
+		 **/
+		template<typename T>
+		inline exact_number<T> secant(exact_number<T> x, size_t max_error_exponent, bool upper){
+			exact_number<T> result("1");
+			exact_number<T> cos = cosine(x, max_error_exponent, upper);
+			result.divide_vector(cos, max_error_exponent, upper);
+			result = result.up_to(max_error_exponent, upper);
+			return result;
+		}

```

Comment:
in fact in this case, too, I would suggest using Taylor expansion

Reply:
Nope, its Taylor expansion is too complex. Calculating it by reciprocating cos(x) is more efficient and easy.
Remember for all terms in Taylor expansion, we need to do division, so dividing it by cos will be equally efficient to cos(x), just one extra division.